stock_picking_group_by_partner_by_carrier/report/report_delivery_slip.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2021 Camptocamp
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+from odoo import api, models
+from odoo.tools import float_round
+
+
+class DeliverySlipReport(models.AbstractModel):
+    _name = "report.stock.report_deliveryslip"
+    _description = "Delivery Slip Report"
+
+    @api.model
+    def get_remaining_to_deliver_data(self, picking):
+        """Return dictionaries encoding pending quantities to deliver
+
+        Returns a list of dictionaries, encoding the data to be displayed
+        at the end of the delivery slip, summarising for each order the
+        pending quantities for each of its lines.
+        """
+        remaining_to_deliver_data = []
+        report_lines = picking.get_delivery_report_lines()
+        for report_line in report_lines:
+            # Depending on the state of the picking, get_delivery_report_lines
+            # may return move.line or stock.move.line, so we check first.
+            move = (
+                report_line.move_id
+                if report_line._name == "stock.move.line"
+                else report_line
+            )
+
+            if not report_line.id:  # Header moves are created as mock moves.
+                remaining_to_deliver_data.append(
+                    {"is_header": True, "concept": report_line.description_picking}
+                )
+            elif (
+                move.picking_id.picking_type_id.code == "outgoing" and move.sale_line_id
+            ):
+                sol = move.sale_line_id
+                picking_state = move.picking_id.state
+                qty = None
+                if picking_state == "done":
+                    qty = sol.product_uom._compute_quantity(
+                        sol.product_uom_qty - sol.qty_delivered, move.product_uom
+                    )
+                elif picking_state not in {"cancel", "done"}:
+                    qty = (
+                        sol.product_uom._compute_quantity(
+                            sol.product_uom_qty - sol.qty_delivered, move.product_uom
+                        )
+                        - move.product_uom_qty
+                    )
+
+                if qty:
+                    remaining_to_deliver_data.append(
+                        {
+                            "is_header": False,
+                            "concept": move.product_id.name_get()[0][-1],
+                            "move": move,
+                            "qty": float_round(
```

Comment:
IMO you should rely on the widget

Reply:
I can't, the widget only considers the precision, and I can't take the general precision for the UoM because here I need the precision, or rounding, for each UoM in particular.